modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -181,6 +181,101 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
   ): F[Either[ResolutionError, Json]] =
     lookupSchemaResult(schemaKey).map(_.map(_.value.schema))
 
+  def lookupSchemasUntil(
+    maxSchemaKey: SchemaKey
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): F[Either[SchemaResolutionError, List[Json]]] = {
+
+    // Looks up addition 0 and keeps looking up until NotFound or other ResolutionError.
+    // At least one schema should get returned. If not, the first ResolutionError gets returned,
+    // even if it is NotFound.
+    def indeterministicFetchAll(
+      schemaKey: SchemaKey
+    ): F[Either[SchemaResolutionError, List[Json]]] = {
+      case class Fetched(
+        error: Option[SchemaResolutionError],
+        schemas: List[Json],
+        finished: Boolean
+      )
+      Monad[F]
+        .iterateWhileM[Fetched](
+          Fetched(None, List.empty[Json], false)
+        ) { case Fetched(error, schemas, finished) =>
+          val nextSchemaKey =
+            schemaKey.copy(version = schemaKey.version.copy(addition = schemas.size))
+          lookupSchema(nextSchemaKey).map {
+            case Left(resolutionError) if resolutionError.isNotFound && schemas.nonEmpty =>
```

Comment:
This line is causing me much anxiety.  The return value of `isNotFound` is only true if **every** Iglu repo returns a 404.  If any single repo returns, 500 then here you return something which causes a failed event.

In our typical setups, we have 1 private iglu server, plus up to 4 mirrors of Iglu Central.  We use 4 mirrors to provide redundancy if any single mirror becomes unreachable.

So imagine a situation where:

- The private repo contains `1-0-0` and `1-1-0` of a schema
- The schema does not exist in Iglu Central, so we always get 404 from Iglu Central
- One of the Iglu Central mirrors has become unhealthy, so it always returns 500

Here are the response codes for each lookup:

| Version | Private Iglu repo | Iglu Central | Iglu Central mirror 1 | Iglu Central mirror 2 |
|---------|-------------------|--------------|-----------------------|-----------------------|
| 1-0-0   | 200               | 404          | 404                   | 500                   |
| 1-0-1   | 404               | 404          | 404                   | 500                   |
| 1-1-0   | 200               | 404          | 404                   | 500                   |

In this scenario, ideally the result should successfully return the `1-0-0` and `1-1-0` schemas.  But your current implementation returns a schema resolution error in this scenario.

I'm still trying to work out what is the best behaviour in this scenario.  Do you have any ideas?

My doubts still are relevant even if you switch to the alternative implementation.  We need to decide whether to check `isNotFound`.

Reply:
What if one of those unhealthy repositories contains `1-0-1`? We never know

Reply:
Interesting idea! Now vendor prefix is only used to prioritize repos, right? We could go even further and don't do lookup for repos without a vendor match (assuming there is a match somewhere else). Like, "don't even try, there is no chance to find anything there" with given schema vendor.

Reply:
> What if one of those unhealthy repositories contains 1-0-1? We never know
 
But we do know that Iglu central and all its mirrors should contain the same schemas, so if one returns 404, they all should.

Reply:
Ideally, we would want that if Iglu Central or one of its mirror returns `404` for a schema, we ignore the `5xx` of the others (not including private repo of course). 

This is what the resolvers looks like for customers.

On GCP:
```
{
  "schema": "iglu:com.snowplowanalytics.iglu/resolver-config/jsonschema/1-0-2",

  "data": {
    "cacheSize": 500,
    "cacheTtl": 3600,
    "repositories": [
      {
        "name": "Iglu Central",
        "priority": 10,
        "vendorPrefixes": [ ],
        "connection": {
          "http": {
            "uri": "http://iglucentral.com"
          }
        }
      },
      {
        "name": "Iglu Central - GCP Mirror",
        "priority": 10,
        "vendorPrefixes": [ ],
        "connection": {
          "http": {
            "uri": "http://mirror01.iglucentral.com"
          }
        }
      },
      {
       ... private repo ...
      }
    ]
  }
}
```

On AWS:
```
{
  "schema": "iglu:com.snowplowanalytics.iglu/resolver-config/jsonschema/1-0-2",
  "data": {
    "cacheSize": 500,
    "cacheTtl": 600,
    "repositories": [
      {
        "name": "Iglu Central",
        "priority": 1,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "http://iglucentral.com"
          }
        }
      },
      {
        "name": "Iglu Central - Mirror 01",
        "priority": 1,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "http://mirror01.iglucentral.com"
          }
        }
      },
      {
        "name": "Iglu Central - Server Mirror eu1",
        "priority": 10,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "https://com-iglucentral-......iglu.snplow.net/api"
          }
        }
      },
      {
        "name": "Iglu Central - Server Mirror eu2",
        "priority": 10,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "https://com-iglucentral-......iglu.snplow.net/api"
          }
        }
      },
      {
      ... private repo ...
      }
    ]
  }
}
```

On Azure:
```
{
  "schema": "iglu:com.snowplowanalytics.iglu/resolver-config/jsonschema/1-0-2",
  "data": {
    "cacheSize": 500,
    "cacheTtl": 600,
    "repositories": [
      {
        "name": "Iglu Central",
        "priority": 1,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "http://iglucentral.com"
          }
        }
      },
      {
        "name": "Iglu Central - Mirror 01",
        "priority": 1,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "http://mirror01.iglucentral.com"
          }
        }
      },
      {
        "name": "Iglu Central - Server Mirror eu1",
        "priority": 10,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "https://com-iglucentral-.....iglu.snplow.net/api"
          }
        }
      },
      {
        "name": "Iglu Central - Server Mirror eu2",
        "priority": 10,
        "vendorPrefixes": [],
        "connection": {
          "http": {
            "uri": "https://com-iglucentral-......iglu.snplow.net/api"
          }
        }
      },
      {
         ... private repo ...
      },
      {
       ... private repo ...
      }
    ]
  }
}
```

In all cases the name contains `Iglu Central` and the `uri` `iglucentral`. We could use that to judge if a repo is Iglu Central or not.

I'm very aware that this is not ideal but in my opinion this is better than returning failed events if a mirror is unhealthy.

The ideal way would be to update the config with a field that says if a repo is Iglu Central or a mirror.

WDYT ?